File: code_samples/006_occupancy_map.cpp
/*

CODE SAMPLE # 006: Occupancy Map
This code sample allows users to access the region map within a depth range.



>>>>>> Compile this code using the following command....

g++  006_occupancy_map.cpp /usr/src/tensorrt/bin/common/logger.o ../lib/libPAL.so ../lib/libPAL_CAMERA.so ../lib/libPAL_DEPTH_HQ.so ../lib/libPAL_DEPTH_128.so  ../lib/libPAL_DE.so ../lib/libPAL_EDET.so `pkg-config --libs --cflags opencv`   -O3  -o  006_occupancy_map.out -I../include/ -lv4l2 -lpthread -lcudart -L/usr/local/cuda/lib64 -lnvinfer -lnvvpi -lnvparsers -lnvinfer_plugin -lnvonnxparser -lmyelin -lnvrtc -lcudart -lcublas -lcudnn -lrt -ldl -lstdc++fs -lX11

>>>>>> Execute the binary file by typing the following command...

./006_occupancy_map.out


>>>>>> KEYBOARD CONTROLS:
	Press ESC to close the window
	Press v/V to toggle vertical flip property
	Press f/F to toggle filter rgb property
	Press d/D to toggle fast depth property
	Press r/R to toggle near range property

*/


# include <stdio.h>
# include <opencv2/opencv.hpp>
# include "PAL.h"
#include <X11/Xlib.h>
using namespace cv;
using namespace std;

namespace PAL
{
	namespace Internal
	{
		void EnableDepth(bool flag);
		void MinimiseCompute(bool flag);
	}
}

cv::Mat Getoccupancy1D(Mat rgb_image, Mat depth_mat, int depth_thresh, float context_threshold) 
{
	cv::Mat mask;
	
	cv::threshold(depth_mat, mask, depth_thresh, 255, cv::THRESH_BINARY_INV);
	mask = mask/255;
	cv::Mat occupancySum = cv::Mat::zeros(rgb_image.rows, rgb_image.cols, CV_32FC1);
	cv::reduce(mask, occupancySum, 0, cv::REDUCE_SUM, CV_32FC1);
	cv::Mat occupancy1D = cv::Mat::zeros(rgb_image.rows, rgb_image.cols, CV_8UC1); //initialize with zeros;
	cv::threshold(occupancySum, occupancy1D, rgb_image.rows*context_threshold/100.0, 255, cv::THRESH_BINARY_INV);
	occupancy1D = occupancy1D/255;
	occupancy1D.convertTo(occupancy1D, CV_8UC1);		

	return occupancy1D;
}

int main(int argc, char *argv[])
{
	namedWindow( "PAL Occupancy Map", WINDOW_NORMAL ); // Create a window for display.

	//Depth should be enable for occupancy map as a prerequisite	
	bool isDepthEnabled = true;
	PAL::Internal::EnableDepth(isDepthEnabled);
	PAL::Internal::MinimiseCompute(false);
	int width, height;
	if(PAL::Init(width, height,-1) != PAL::SUCCESS) //Connect to the PAL camera
	{
		printf("Camera Init failed\n");
		return 1;
	}

	PAL::CameraProperties data; 
	PAL::Acknowledgement ack = PAL::LoadProperties("../Explorer/SavedPalProperties.txt", &data);
	if(ack != PAL::SUCCESS)
	{
	    printf("Error Loading settings\n");
	}

	PAL::CameraProperties prop;
	
	unsigned int flag = PAL::MODE | PAL::FD | PAL::NR | PAL::FILTER_SPOTS | PAL::VERTICAL_FLIP;
	prop.mode = PAL::Mode::POINT_CLOUD_25D; // The other available option is PAL::Mode::HIGH_QUALITY_DEPTH
	prop.fd = 1;
	prop.nr = 1;
	prop.filter_spots = 1;
	prop.vertical_flip =0;
	PAL::SetCameraProperties(&prop, &flag);

	//depth threshold in cm
	//The depth threshold should be kept within 1m to 2m range.
	int threshold_cm = (argc>1) ? atof(argv[1]) : 100;

	//context threshold in percentage to be considered for occupancy
	//The context threshold should be kept within 50(recommended) to 80 range.
	int context_threshold = (argc>2) ? atof(argv[2]) : 50;	

	if (threshold_cm > 200)
	{
		threshold_cm = 200;
		printf("depth threshold set above maximum range. Setting to 2m\n");
	}
	else if (threshold_cm < 100)
	{
		threshold_cm = 100;
		printf("depth threshold set below minimum range. Setting to 1m\n");
	}

	if (context_threshold > 80)
	{
		context_threshold = 80;
		printf("context threshold set above maximum range. Setting to 80\n");
	}
	else if(context_threshold < 50)
	{
		context_threshold = 50;
		printf("context threshold set below miminum range. Setting to 50\n");
	}

	//width and height are the dimensions of each panorama.
	//Each of the panoramas are displayed at quarter their original resolution.
	//Since the left+right+disparity are vertically stacked, the window height should be thrice the quarter-height
	// Getting Screen resolution 
	Display* disp = XOpenDisplay(NULL);
	Screen*  scrn = DefaultScreenOfDisplay(disp);
	int sc_height = scrn->height;
	int sc_width  = scrn->width;
	
	resizeWindow("PAL Occupancy Map", sc_width, sc_height);//width/4, (height/4)*2);

	int key = ' ';

	printf("Press ESC to close the window\n");    
	printf("Press v/V to toggle vertical flip property\n");
	printf("Press f/F to toggle filter rgb property\n");
	printf("Press r/R to toggle near range property\n");
	
	size_t currentResolution = 0;
	//std::vector<PAL::Resolution> resolutions = PAL::GetAvailableResolutions();
	bool filter_spots = true;	
	bool flip = false;
	bool nr = true;
	bool fd = true;

	//27 = esc key. Run the loop until the ESC key is pressed
	
	while(key != 27)
   	{
		cv::Mat rgb, depth, output, right;
		PAL::Image left_img, right_img, depth_img, disparity;
		
		PAL::GrabFrames(&left_img, &right_img, &depth_img);
		
		depth = cv::Mat(depth_img.rows, depth_img.cols, CV_32FC1, depth_img.Raw.f32_data);
		rgb = cv::Mat(left_img.rows, left_img.cols, CV_8UC3, left_img.Raw.u8_data);
		cv::resize(depth, depth , cv::Size(rgb.cols, rgb.rows));
		
		PAL::CameraProperties data;
		PAL::GetCameraProperties(&data);
		
		
        if(isDepthEnabled)      
        {

			depth.convertTo(depth, CV_8UC1);
			
			cv::Mat occupancy1D = Getoccupancy1D(rgb, depth,threshold_cm,context_threshold);

			cv::Mat r = cv::Mat::ones(1, rgb.cols, CV_8UC1);
 			cv::Mat g = cv::Mat::ones(1, rgb.cols, CV_8UC1);
			cv::Mat b = cv::Mat::ones(1, rgb.cols, CV_8UC1);

  			unsigned char* poccupancy1D = (unsigned char* )occupancy1D.data;
  			unsigned char* pr = (unsigned char* )r.data;

			for(int i=0; i<rgb.cols; i++,poccupancy1D++,pr++)
			{
				if((*poccupancy1D))
				{
					*pr = (*pr)*2;
				}
			}
			
			cv::Mat final_img;
			vector<Mat> channels;
			channels.push_back(b);
			channels.push_back(g);
			channels.push_back(r);

			merge(channels, final_img);
			resize(final_img, final_img, rgb.size());
 			cv::Mat colored_out = rgb.mul(final_img);
			imshow("PAL Occupancy Map", colored_out);//colored_out);
        }
        else
        {	
            //Display the final rgb image
            imshow( "PAL Occupancy Map", rgb); 
        }		

		//Wait for the keypress - with a timeout of 1 ms
		key = waitKey(1) & 255;


		if (key == 'f' || key == 'F')
		{	
			PAL::CameraProperties prop;
			filter_spots = !filter_spots;
			prop.filter_spots = filter_spots;
			unsigned int flags = PAL::FILTER_SPOTS;
			PAL::SetCameraProperties(&prop, &flags);
		}
		if (key == 'v' || key == 'V')
		{		    
			PAL::CameraProperties prop;
			flip = !flip;
			prop.vertical_flip = flip;
			unsigned int flags = PAL::VERTICAL_FLIP;
			PAL::SetCameraProperties(&prop, &flags);
		}
		
		if(key == 'r' || key == 'R')
		{		
			PAL::CameraProperties prop;
			nr = !nr;
			prop.nr = nr;
			unsigned int flags = PAL::NR;
			PAL::SetCameraProperties(&prop, &flags);
		}
    }

    printf("exiting the application\n");
    PAL::Destroy();
    
    return 0;
}

File: code_samples/python/006_occupancy_map.py
# CODE SAMPLE # 006: Occupancy Map
# This code sample allows users to access the region map within a depth range.

import sys
import PAL_PYTHON
import cv2
import numpy as np

def GetUserInput(prompt):
    while True:
        try:
            value = int(input(prompt))
            return value
        except ValueError:
            print("Invalid input. Please enter an integer.")

def Getoccupancy1D(rgb_image, depth_mat, depth_thresh, context_threshold):
    ret, mask = cv2.threshold(depth_mat, depth_thresh, 255, cv2.THRESH_BINARY_INV)
    occupancySum = np.sum(mask.astype(np.float32), axis=0) / 255. 
    ret, occupancy1D = cv2.threshold(occupancySum, rgb_image.shape[0]*context_threshold/100.0, 255, cv2.THRESH_BINARY_INV);
    return (occupancy1D/255).astype(np.uint8).reshape(1,-1)

def main():
    #depth threshold in cm
    #The depth threshold should be kept within 1m to 2m range.
    threshold_cm = GetUserInput("Enter the depth threshold in cm, eg 100\n")
    if(threshold_cm > 200):
        threshold_cm = 200
        print("depth threshold set above maximum range. Setting to 2m")
    elif(threshold_cm < 100):
        threshold_cm = 100
        print("depth threshold set below minimum range. Setting to 1m")

    #context threshold in percentage to be considered for occupancy
    #The context threshold should be kept within 50(recommended) to 80 range.
    context_threshold = GetUserInput("Enter the Context threshold. It is the percentage to be considered for occupancy, eg 50\n")
    if(context_threshold > 80):
        context_threshold = 80
        print("context threshold set above maximum range. Setting to 80")
    elif(context_threshold < 50):
        context_threshold = 50
        print("context threshold set below miminum range. Setting to 50")

    #Camera index is the video index assigned by the system to the camera. 
    #By default we set it to 5. Specify the index if the value has been changed.
    camera_index = 5    
    
    arg = len(sys.argv)
    if arg == 2:
        camera_index = int(sys.argv[1])

    #Connect to the PAL camera    
    res_init = PAL_PYTHON.InitP(camera_index)

    if res_init != PAL_PYTHON.SUCCESSP:
        print("Camera Init failed\n")
        return

    #Setting API Mode
    PAL_PYTHON.SetAPIModeP(PAL_PYTHON.DEPTHP)

    loaded_prop = {}
    loaded_prop = PAL_PYTHON.createPALCameraPropertiesP(loaded_prop)
    
    #Loading camera properties from a text file
    loaded_prop, ack_load = PAL_PYTHON.LoadPropertiesP("../../Explorer/SavedPalProperties.txt", loaded_prop)
    if ack_load == PAL_PYTHON.INVALID_PROPERTY_VALUEP: 
        PAL_PYTHON.DestroyP()
        return
    if ack_load != PAL_PYTHON.SUCCESSP:
        print("Error Loading settings! Loading default values.")

    # Creating a window
    source_window = 'PAL Occupancy Map'
    cv2.namedWindow(source_window, cv2.WINDOW_AUTOSIZE)
    
    print("\n\nPress ESC to close the window.")
    print("Press v/V to flip vertically.")    
    print("Press f/F to toggle filter rgb property.\n\n")

    key = ' '
    
    # ESC
    while key != 27:
        # GrabFrames function
        left, right, depth, raw_depth, camera_changed  = PAL_PYTHON.GrabDepthDataP()
        if camera_changed == True:
            break

        # FLOAT->RGB
        left_mat = left
        if bool(loaded_prop["raw_depth"]):
            depth_mat = raw_depth
        else:
            depth_mat = depth
        
        occupancy1D = Getoccupancy1D(left_mat, depth_mat, threshold_cm, context_threshold)
        mask  = cv2.resize(occupancy1D, (left_mat.shape[1], left_mat.shape[0]), interpolation =cv2.INTER_NEAREST) * 255
        
        color_mask = left_mat.copy();
        color_mask[mask == 0] = (0,0,255)
        left_mat = (left_mat.astype(np.float32) * 0.7 + color_mask.astype(np.float32) * 0.3).astype(np.uint8) 
        
        # Show results        
        cv2.imshow(source_window, left_mat)

        # Wait for 1ms
        key = cv2.waitKey(1) & 255

        if key == 102:            
            flag = PAL_PYTHON.FILTER_SPOTSP
            loaded_prop["filter_spots"] = not(bool(loaded_prop["filter_spots"]))
            loaded_prop, flags, res_scp = PAL_PYTHON.SetCameraPropertiesP(loaded_prop, flag)

        if key == 118:            
            flag = PAL_PYTHON.VERTICAL_FLIPP
            loaded_prop["vertical_flip"] = not(bool(loaded_prop["vertical_flip"]))
            loaded_prop, flags, res_scp = PAL_PYTHON.SetCameraPropertiesP(loaded_prop, flag)

    # Destroying connections
    print("exiting the application\n")
    PAL_PYTHON.DestroyP()

    return

if __name__ == "__main__":
    main()
